contracts/beyondFramework/dca/base/DCABaseUpgradeable.sol

```diff
@@ -0,0 +1,1168 @@
+    }
+
+    modifier emergencyWithdrawOnEmergencyExitedStatus(address sender) {
+        // if emergency exited then user should receive everything in deposit asset
+        if (isEmergencyExited()) {
+            _emergencyWithdrawUserDeposit(sender);
+            return;
+        }
+
+        _;
+    }
+
+    receive() external payable {}
+
+    // ----- Base Class Methods -----
+    function deposit(uint256 amount, uint8 amountSplit)
+        public
+        virtual
+        nonReentrant
+        whenNotPaused
+        nonEmergencyExited
+    {
+        _deposit(_msgSender(), amount, amountSplit);
+    }
+
+    function depositFor(
+        address sender,
+        uint256 amount,
+        uint8 amountSplit
+    )
+        public
+        virtual
+        onlyPortfolio
+        nonReentrant
+        whenNotPaused
+        nonEmergencyExited
+    {
+        _deposit(sender, amount, amountSplit);
+    }
+
+    function _deposit(
+        address sender,
+        uint256 amount,
+        uint8 amountSplit
+    ) private {
+        // assert valid amount sent
+        if (amount < minDepositAmount) {
+            revert TooSmallDeposit();
+        }
+
+        // transfer deposit token from portfolio
+        depositTokenInfo.token.safeTransferFrom(
+            _msgSender(),
+            address(this),
+            amount
+        );
+
+        // compute actual deposit and transfer fee to receiver
+        amount = _takeFee(amount);
+
+        DcaDepositor storage depositor = depositors[sender];
+
+        // assert positions limit is not reached
+        if (depositor.positions.length == positionsLimit) {
+            revert PositionsLimitReached();
+        }
+
+        // add splitted amounts to the queue
+        globalInvestQueue.splitUserInvestmentAmount(amount, amountSplit);
+
+        // if not started position with the same split exists - increase deposit amount
+        for (uint256 i = 0; i < depositor.positions.length; i++) {
+            // calculate amount of passed investment epochs
+            uint256 passedInvestPeriods = (lastInvestmentTimestamp -
+                depositor.positions[i].investedAt) / investmentPeriod;
+
+            if (
+                passedInvestPeriods == 0 &&
+                depositor.positions[i].amountSplit == amountSplit
+            ) {
+                // not started position with the same amount split exists
+                // just add invested amount here
+                depositor.positions[i].depositAmount += amount;
+                return;
+            }
+        }
+
+        // otherwise create new position
+        depositor.positions.push(
+            Position(
+                amount,
+                amountSplit,
+                lastInvestmentTimestamp,
+                dcaHistory.currentHistoricalIndex()
+            )
+        );
+
+        emit Deposit(sender, amount, amountSplit);
+    }
+
+    function invest()
+        public
+        virtual
+        onlyDcaInvestor
```

Comment:
is there any reason to protect the call to the invest method by onlyDcaInvestor?

Reply:
probably we should do the same in processReward, I wonder why this one was not picked up by dedaub, but all the other unimportant stuff were